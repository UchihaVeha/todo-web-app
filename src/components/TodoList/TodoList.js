// @flow
import React from 'react';
import { compose, pure } from 'recompose';
import { withStyles } from 'material-ui/styles';
import { List, Paper } from 'material-ui';
import { CircularProgress } from 'material-ui/Progress';
import type { TodoIds } from 'modules/todo/reducers';

const styles = theme => ({
  list: {
    display: 'flex',
    flexDirection: 'column',
    paddingTop: theme.spacing.unit,
    background: 'white'
  },
  centered: {
    textAlign: 'center'
  }
});

type Props = {
  todoIds: TodoIds,
  isLoading: boolean,
  classes: Object,
  listItem: any
};
const TodoList = ({
  todoIds,
  isLoading,
  classes,
  listItem: ListItem
}: Props): * => (
  <div>
    <Paper className={classes.list}>
      <div className={classes.centered}>
        {!isLoading && todoIds.count() === 0 && <span>No items</span>}
        {isLoading && <CircularProgress size={30} />}
      </div>
      <List>{todoIds.map(id => <ListItem key={id} id={id} />).toArray()}</List>
    </Paper>
  </div>
);

export default compose(pure, withStyles(styles, { name: 'TodoList' }))(
  TodoList
);
