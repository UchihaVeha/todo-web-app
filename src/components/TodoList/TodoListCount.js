// @flow
import React from 'react';
import { compose, pure, withHandlers } from 'recompose';
import { withRouter } from 'react-router';
import { withStyles } from 'material-ui/styles';
import { Paper, Button, Typography } from 'material-ui';
import AddIcon from 'material-ui-icons/Add';
import { grey } from 'material-ui/colors';

const styles = theme => ({
  total: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: theme.spacing.unit,
    color: theme.palette.common.darkWhite,
    background: grey['700'],
    boxShadow: `inset ${theme.shadows[6]}`
  },
  label: {
    fontWeight: 900
  },
  button: {
    zIndex: 2000,
    alignSelf: 'flex-end',
    marginTop: '-36px'
  }
});

type Props = {
  count: number,
  classes: Object,
  history: Object,
  handleCreateClick: number => void
};
const TodoListCount = ({ count, classes, handleCreateClick }: Props): * => (
  <Paper square>
    <div className={classes.total}>
      <Typography type="caption" color="inherit" className={classes.label}>
        TOTAL: {count}
      </Typography>
      <Button
        fab
        color="accent"
        className={classes.button}
        onClick={handleCreateClick}
      >
        <AddIcon />
      </Button>
    </div>
  </Paper>
);

export default compose(
  pure,
  withRouter,
  withHandlers({
    handleCreateClick: props => () => {
      props.history.push('/create');
    }
  }),
  withStyles(styles, { name: 'TodoListCount' })
)(TodoListCount);
