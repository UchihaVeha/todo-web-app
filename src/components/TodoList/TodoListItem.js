'no babel-plugin-flow-react-proptypes';

// @flow
import React from 'react';
import { compose, pure, withHandlers } from 'recompose';
import { withRouter } from 'react-router';
import { withStyles } from 'material-ui/styles';
import {
  ListItem,
  ListItemText,
  ListItemSecondaryAction
} from 'material-ui/List';
import { Checkbox, IconButton } from 'material-ui';
import EditIcon from 'material-ui-icons/Edit';
import DeleteIcon from 'material-ui-icons/Delete';
import Menu, { MenuItem } from 'material-ui/Menu';
import MoreVertIcon from 'material-ui-icons/MoreVert';
import type { TodoRecord } from 'modules/todo/reducers';

const styles = () => ({
  root: {
    display: 'none'
  }
});

type Props = {
  todo: TodoRecord,
  tagNames: *,
  isHidden: boolean,
  isPending: boolean,
  handleOnClick: number => void,
  handleEditClick: number => void,
  handleDeleteClick: number => void,
  classes: Object
};

class TodoListItemMenu extends React.Component {
  state = {
    anchorEl: null,
    open: false
  };

  handleClick = event => {
    this.setState({ open: true, anchorEl: event.currentTarget });
  };

  handleRequestClose = () => {
    this.setState({ open: false });
  };

  render() {
    const { handleEditClick, handleDeleteClick } = this.props;
    return (
      <div>
        <IconButton
          aria-label="More"
          aria-owns={this.state.open ? 'todo-list-item-menu' : null}
          aria-haspopup="true"
          onClick={this.handleClick}
        >
          <MoreVertIcon />
        </IconButton>
        <Menu
          id="todo-list-item-menu"
          anchorEl={this.state.anchorEl}
          open={this.state.open}
          onRequestClose={this.handleRequestClose}
        >
          <MenuItem key={0} onClick={this.handleRequestClose}>
            <IconButton aria-label="Edit" onClick={handleEditClick}>
              <EditIcon />
            </IconButton>
          </MenuItem>
          <MenuItem key={1} onClick={this.handleRequestClose}>
            <IconButton aria-label="Delete" onClick={handleDeleteClick}>
              <DeleteIcon />
            </IconButton>
          </MenuItem>
        </Menu>
      </div>
    );
  }
}

const TodoListItem = ({
  todo,
  tagNames,
  isHidden,
  isPending,
  handleOnClick,
  handleEditClick,
  handleDeleteClick,
  classes
}: Props) => (
  <ListItem
    button
    divider
    disableGutters
    disabled={isPending}
    onClick={handleOnClick}
    classes={isHidden ? classes : {}}
  >
    <Checkbox disableRipple tabIndex="-1" checked={todo.isCompleted} />
    <ListItemText
      primary={todo.title}
      secondary={tagNames.map(tag => <span key={tag}>{`${tag} `}</span>)}
    />
    <ListItemSecondaryAction>
      <TodoListItemMenu
        handleEditClick={handleEditClick}
        handleDeleteClick={handleDeleteClick}
      />
    </ListItemSecondaryAction>
  </ListItem>
);

export default compose(
  pure,
  withRouter,
  withHandlers({
    handleOnClick: props => () => {
      props.onToggleIsCompleted(props.todo, props.tagNames);
    },
    handleEditClick: props => () => {
      props.history.push(`/update/${props.todo.id}`);
    },
    handleDeleteClick: props => () => {
      props.onDeleteTodo(props.todo);
    }
  }),
  withStyles(styles, { name: 'TodoListItem' })
)(TodoListItem);
