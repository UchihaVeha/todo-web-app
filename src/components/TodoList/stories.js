import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { OrderedSet } from 'immutable';
import { MemoryRouter } from 'react-router-dom';
import TodoList from './TodoList';
import TodoListItem from './TodoListItem';

storiesOf('Todo List Item', module)
  .addDecorator(story => (
    <div style={{ padding: 30, maxWidth: 900 }}>
      <MemoryRouter>{story()}</MemoryRouter>
    </div>
  ))
  .add('is not completed', () => (
    <TodoListItem
      todo={{
        id: 1,
        title: 'Todo',
        isCompleted: false
      }}
      tagNames={[]}
      isHidden={false}
      isPending={false}
      onToggleIsCompleted={action('on click')}
      onDeleteTodo={action('delete click')}
    />
  ))
  .add('is completed', () => (
    <TodoListItem
      todo={{
        id: 1,
        title: 'Todo',
        isCompleted: true
      }}
      tagNames={[]}
      isHidden={false}
      isPending={false}
      onToggleIsCompleted={action('on click')}
      onDeleteTodo={action('delete click')}
    />
  ))
  .add('with tags', () => (
    <TodoListItem
      todo={{
        id: 1,
        title: 'Todo',
        isCompleted: false
      }}
      tagNames={['React', 'Angular']}
      isHidden={false}
      isPending={false}
      onToggleIsCompleted={action('on click')}
      onDeleteTodo={action('delete click')}
    />
  ))
  .add('is pending', () => (
    <TodoListItem
      todo={{
        id: 1,
        title: 'Todo',
        isCompleted: false
      }}
      tagNames={['React', 'Angular']}
      isHidden={false}
      isPending
      onToggleIsCompleted={action('on click')}
      onDeleteTodo={action('delete click')}
    />
  ));

storiesOf('Todo List', module)
  .addDecorator(story => (
    <div style={{ padding: 30, maxWidth: 600 }}>
      <MemoryRouter>{story()}</MemoryRouter>
    </div>
  ))
  .add('default', () => (
    <TodoList
      todoIds={OrderedSet([1, 2])}
      isLoading={false}
      listItem={() => (
        <TodoListItem
          todo={{
            id: 1,
            title: 'Todo',
            isCompleted: false
          }}
          tagNames={[]}
          isHidden={false}
          isPending={false}
          onToggleIsCompleted={action('on click')}
          onDeleteTodo={action('delete click')}
        />
      )}
    />
  ))
  .add('is loading', () => <TodoList todoIds={OrderedSet([])} isLoading />)
  .add('empty', () => <TodoList todoIds={OrderedSet([])} isLoading={false} />);
