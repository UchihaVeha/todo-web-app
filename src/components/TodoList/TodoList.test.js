import React from 'react';
import { shallow } from 'enzyme';
import { OrderedMap } from 'immutable';
import TodoList from './TodoList';

describe('TodoList component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<TodoList todoIds={OrderedMap([])} isLoading={false} />);
  });

  it('should render pure', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render withStyles', () => {
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('should render component', () => {
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });
});
