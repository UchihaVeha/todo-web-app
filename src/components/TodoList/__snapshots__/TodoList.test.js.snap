// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should render component 1`] = `
<div>
  <withStyles(Paper)
    className="TodoList-list-1"
  >
    <div
      className="TodoList-centered-2"
    >
      <span>
        No items
      </span>
    </div>
    <withStyles(List) />
  </withStyles(Paper)>
</div>
`;

exports[`TodoList component should render pure 1`] = `
<withStyles(TodoList)
  isLoading={false}
  todoIds={Immutable.OrderedMap {}}
/>
`;

exports[`TodoList component should render withStyles 1`] = `
<TodoList
  classes={
    Object {
      "centered": "TodoList-centered-2",
      "list": "TodoList-list-1",
    }
  }
  isLoading={false}
  todoIds={Immutable.OrderedMap {}}
/>
`;
